ui/article-aem-renderer/src/main/java/org/cru/godtools/article/aem/service/AemArticleManager.kt

```diff
@@ -0,0 +1,316 @@
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.actor
+import kotlinx.coroutines.channels.receiveOrNull
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.conflate
+import kotlinx.coroutines.joinAll
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.withTimeoutOrNull
+import org.ccci.gto.android.common.base.TimeConstants.HOUR_IN_MS
+import org.ccci.gto.android.common.base.TimeConstants.MIN_IN_MS
+import org.ccci.gto.android.common.db.Query
+import org.ccci.gto.android.common.kotlin.coroutines.MutexMap
+import org.ccci.gto.android.common.kotlin.coroutines.ReadWriteMutex
+import org.ccci.gto.android.common.kotlin.coroutines.withLock
+import org.cru.godtools.article.aem.api.AemApi
+import org.cru.godtools.article.aem.db.ArticleRoomDatabase
+import org.cru.godtools.article.aem.model.AemImport
+import org.cru.godtools.article.aem.model.Resource
+import org.cru.godtools.article.aem.service.support.extractResources
+import org.cru.godtools.article.aem.service.support.findAemArticles
+import org.cru.godtools.article.aem.util.AemFileManager
+import org.cru.godtools.article.aem.util.addExtension
+import org.cru.godtools.base.tool.service.ManifestManager
+import org.cru.godtools.model.Tool
+import org.cru.godtools.model.Translation
+import org.cru.godtools.xml.model.Manifest
+import org.keynote.godtools.android.db.Contract.ToolTable
+import org.keynote.godtools.android.db.Contract.TranslationTable
+import org.keynote.godtools.android.db.GodToolsDao
+import timber.log.Timber
+
+private const val TAG = "AemArticleManager"
+
+@VisibleForTesting
+internal const val CLEANUP_DELAY = HOUR_IN_MS
+@VisibleForTesting
+internal const val CLEANUP_DELAY_INITIAL = MIN_IN_MS
+private const val CACHE_BUSTING_INTERVAL_JSON = HOUR_IN_MS
+
+@VisibleForTesting
+internal val QUERY_ARTICLE_TRANSLATIONS = Query.select<Translation>()
+    .join(TranslationTable.SQL_JOIN_TOOL)
+    .where(ToolTable.FIELD_TYPE.eq(Tool.Type.ARTICLE).and(TranslationTable.SQL_WHERE_DOWNLOADED))
+
+@Singleton
+class AemArticleManager @VisibleForTesting internal constructor(
+    private val aemDb: ArticleRoomDatabase,
+    private val api: AemApi,
+    private val dao: GodToolsDao,
+    private val fileManager: AemFileManager,
+    private val manifestManager: ManifestManager,
+    private val coroutineScope: CoroutineScope
+) {
+    @Inject
+    internal constructor(
+        aemDb: ArticleRoomDatabase,
+        api: AemApi,
+        dao: GodToolsDao,
+        fileManager: AemFileManager,
+        manifestManager: ManifestManager
+    ) : this(aemDb, api, dao, fileManager, manifestManager, CoroutineScope(Dispatchers.Default + SupervisorJob()))
```

Comment:
building out the coroutine scope you can provide several different types of context to it. What this is doing is providing the following 2 pieces of context:
- the default dispatcher
   - coroutines in this scope will execute on this dispatcher by default
- a Supervisor Job
   - This provides a parent job that all other jobs launched from this scope will be children of
   - If you cancel the parent job, it will automatically cancel any descendant jobs that are still running